File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int total = 0;
        for (int n : nums)
            total += n;
        if (total % 2 == 1)
            return false;
        int target = total / 2;
        vector<int> dp(target+1,false);
        dp[0] = true;
        for (int x : nums){
            for (int s=target; s>=x; s--)
                dp[s] = dp[s] || dp[s-x];
        }
        return dp[target];
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        target = total // 2
        if total % 2 != 0:
            return False
        print("target:",target)
        dp = [False] * (target + 1)
        dp[0] = True
        for x in nums:
            for s in range(target,x-1,-1):
                dp[s] = dp[s] or dp[s-x]
                if s == 11:
                    print("now:",dp[s])
        return dp[target]